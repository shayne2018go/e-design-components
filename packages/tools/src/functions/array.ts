export interface IteratorFunctionVoid<T> {
  (item: T, index: number, collection: T[]): void
  (item: T, index: number, collection: T[]): T
  (item: T, index: number, collection: T[]): boolean
}
function forEach<T>(callback: IteratorFunctionVoid<T>, collection: T[]): void
function forEach<T>(callback: IteratorFunctionVoid<T>, collection: T[]) {}

export const treeToArray = () => {}

export const arrayToTree = <T>(
  array: T[],
  opt: any = {
    parentKey: "parentId",
    key: "id",
    childrenKey: "children",
    rootId: undefined,
  }
) => {
  const { key, parentKey, childrenKey, rootId } = opt

  const tree = [] as T[] // 存放结果集

  const itemMap = {} //
  for (const item of array) {
    const id = item[key]
    const pid = item[parentKey]

    if (!itemMap[id]) {
      itemMap[id] = {
        [childrenKey]: [],
      }
    }

    itemMap[id] = {
      ...item,
      [childrenKey]: itemMap[id][childrenKey],
    }

    const treeItem = itemMap[id]

    if (pid === rootId) {
      tree.push(treeItem)
    } else {
      if (!itemMap[pid]) {
        itemMap[pid] = {
          [childrenKey]: [],
        }
      }
      itemMap[pid][childrenKey].push(treeItem)
    }
  }
  return tree
}

export const baseEachTree = (
  iteratee: (node: any, parent: any, deep: number, index: number, iterable?: any[]) => any,
  tree: any[],
  newTree: any[],
  parent = null,
  opt: any = {
    parentKey: "parentId",
    key: "id",
    childrenKey: "children",
    deep: 1,
  }
) => {
  const { childrenKey = "children", deep = 1 } = opt || {}

  const iterable = Object(tree)
  const length = iterable.length

  let index = -1

  while (++index < length) {
    const iterableItem = iterable[index]

    let iterableItemTemp = iteratee(iterableItem, parent, deep, index, iterable) ?? iterableItem
    if (iterableItemTemp === false) {
      break
    }

    const newIterable = iterableItem[childrenKey]
    if (newIterable && newIterable.length) {
      iterableItemTemp[childrenKey] = []
      opt.deep = deep + 1

      baseEachTree(iteratee, newIterable, iterableItemTemp[childrenKey], iterableItemTemp, opt)
    }

    newTree.push(iterableItemTemp)
  }
}

export const eachTree = <T>(fn: (node: T, parent: T, deep: number, index: number) => any, tree: T[], opt?: any) => {
  baseEachTree(fn, tree, [], null, opt)
}

export const mapTree = <T>(fn: (node: T, parent: T, deep: number, index: number) => any, tree: T[], opt?: any) => {
  const newTree = []
  baseEachTree(fn, tree, newTree, null, opt)
  return newTree
}
