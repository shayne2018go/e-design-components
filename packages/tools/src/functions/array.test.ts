import { describe, expect, test, it } from "vitest"
import { arrayToTree, eachTree, mapTree } from "./array"

const _array = [
  { id: 1, name: "部门1", pid: 0 },
  { id: 2, name: "部门2", pid: 1 },
  { id: 3, name: "部门3", pid: 1 },
  { id: 4, name: "部门4", pid: 3 },
  { id: 5, name: "部门5", pid: 4 },
]

const _tree = [
  {
    children: [
      {
        children: [],
        id: 2,
        name: "部门2",
        pid: 1,
      },
      {
        children: [
          {
            children: [
              {
                children: [],
                id: 5,
                name: "部门5",
                pid: 4,
              },
            ],
            id: 4,
            name: "部门4",
            pid: 3,
          },
        ],
        id: 3,
        name: "部门3",
        pid: 1,
      },
    ],
    id: 1,
    name: "部门1",
    pid: 0,
  },
]

test("arrayToTree", () => {
  const res = arrayToTree(_array, {
    key: "id",
    childrenKey: "children",
    parentKey: "pid",
    rootId: 0,
  })

  expect(res).toMatchSnapshot(_tree)
})

test("eachTree", () => {
  const res = [] as any[]
  eachTree(
    (node, parent, deep, index) => {
      res.push({ node, parent, deep, index })
    },
    _tree,
    { childrenKey: "children" }
  )
  expect(res.length).toBe(5)

  expect(res).toMatchSnapshot()
})

test("mapTree", () => {
  const res = mapTree(
    (node, parent, deep, index) => {
      node.name = node.name + "哈哈哈"
      return node
    },
    _tree,
    { childrenKey: "children" }
  )

  expect(res.length).toBe(1)

  expect(res).toMatchSnapshot()
})
