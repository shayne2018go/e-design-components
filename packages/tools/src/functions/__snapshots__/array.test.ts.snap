// Vitest Snapshot v1

exports[`arrayToTree 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": 2,
        "name": "部门2",
        "pid": 1,
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "id": 5,
                "name": "部门5",
                "pid": 4,
              },
            ],
            "id": 4,
            "name": "部门4",
            "pid": 3,
          },
        ],
        "id": 3,
        "name": "部门3",
        "pid": 1,
      },
    ],
    "id": 1,
    "name": "部门1",
    "pid": 0,
  },
]
`;

exports[`eachTree 1`] = `
[
  {
    "deep": 1,
    "index": 0,
    "node": {
      "children": [
        {
          "children": [],
          "id": 2,
          "name": "部门2",
          "pid": 1,
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "id": 5,
                  "name": "部门5",
                  "pid": 4,
                },
              ],
              "id": 4,
              "name": "部门4",
              "pid": 3,
            },
          ],
          "id": 3,
          "name": "部门3",
          "pid": 1,
        },
      ],
      "id": 1,
      "name": "部门1",
      "pid": 0,
    },
    "parent": null,
  },
  {
    "deep": 2,
    "index": 0,
    "node": {
      "children": [],
      "id": 2,
      "name": "部门2",
      "pid": 1,
    },
    "parent": {
      "children": [
        {
          "children": [],
          "id": 2,
          "name": "部门2",
          "pid": 1,
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "id": 5,
                  "name": "部门5",
                  "pid": 4,
                },
              ],
              "id": 4,
              "name": "部门4",
              "pid": 3,
            },
          ],
          "id": 3,
          "name": "部门3",
          "pid": 1,
        },
      ],
      "id": 1,
      "name": "部门1",
      "pid": 0,
    },
  },
  {
    "deep": 2,
    "index": 1,
    "node": {
      "children": [
        {
          "children": [
            {
              "children": [],
              "id": 5,
              "name": "部门5",
              "pid": 4,
            },
          ],
          "id": 4,
          "name": "部门4",
          "pid": 3,
        },
      ],
      "id": 3,
      "name": "部门3",
      "pid": 1,
    },
    "parent": {
      "children": [
        {
          "children": [],
          "id": 2,
          "name": "部门2",
          "pid": 1,
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "id": 5,
                  "name": "部门5",
                  "pid": 4,
                },
              ],
              "id": 4,
              "name": "部门4",
              "pid": 3,
            },
          ],
          "id": 3,
          "name": "部门3",
          "pid": 1,
        },
      ],
      "id": 1,
      "name": "部门1",
      "pid": 0,
    },
  },
  {
    "deep": 3,
    "index": 0,
    "node": {
      "children": [
        {
          "children": [],
          "id": 5,
          "name": "部门5",
          "pid": 4,
        },
      ],
      "id": 4,
      "name": "部门4",
      "pid": 3,
    },
    "parent": {
      "children": [
        {
          "children": [
            {
              "children": [],
              "id": 5,
              "name": "部门5",
              "pid": 4,
            },
          ],
          "id": 4,
          "name": "部门4",
          "pid": 3,
        },
      ],
      "id": 3,
      "name": "部门3",
      "pid": 1,
    },
  },
  {
    "deep": 4,
    "index": 0,
    "node": {
      "children": [],
      "id": 5,
      "name": "部门5",
      "pid": 4,
    },
    "parent": {
      "children": [
        {
          "children": [],
          "id": 5,
          "name": "部门5",
          "pid": 4,
        },
      ],
      "id": 4,
      "name": "部门4",
      "pid": 3,
    },
  },
]
`;

exports[`mapTree 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "id": 2,
        "name": "部门2哈哈哈",
        "pid": 1,
      },
      {
        "children": [
          {
            "children": [
              {
                "children": [],
                "id": 5,
                "name": "部门5哈哈哈",
                "pid": 4,
              },
            ],
            "id": 4,
            "name": "部门4哈哈哈",
            "pid": 3,
          },
        ],
        "id": 3,
        "name": "部门3哈哈哈",
        "pid": 1,
      },
    ],
    "id": 1,
    "name": "部门1哈哈哈",
    "pid": 0,
  },
]
`;
